File: packages/core/src/taxonomy.ts
/** Rule-based hot-path classification (D1). Model is off-path and escalate-only. */

import type {
  ActionRef,
  ActionSignature,
  CaptureFidelity,
  ClassificationRecord,
  ClassificationSource,
  CompensatorCandidate,
  JsonValue,
  PreimageRef,
  Tier,
  TierEvidence,
} from "./types.ts";
import { classifySql } from "./sqlkind.ts";

export const TIER_ORDER: Record<Tier, number> = { T1: 1, T2: 2, T3: 3, T4: 4 };

export function maxTier(a: Tier, b: Tier): Tier {
  return TIER_ORDER[a] >= TIER_ORDER[b] ? a : b;
}

export function minTier(a: Tier, b: Tier): Tier {
  return TIER_ORDER[a] <= TIER_ORDER[b] ? a : b;
}

export interface ClassifyContext {
  result?: JsonValue;
  resultStatus?: number;
  resultHeaders?: Record<string, string>;
  preimage?: PreimageRef;
  compensatorMatched?: boolean;
  manifest?: ActionSignature;
  writableRoots?: string[];
  internalHosts?: string[];
  transportError?: boolean;
  timeout?: boolean;
  recreateFromPreimage?: boolean;
  pkResolvable?: boolean;
  affectedRows?: number;
  maxPreimageBytes?: number;
  maxPreimageRows?: number;
  networkOrFuseMount?: boolean;
  realpathEscapes?: boolean;
  candidates?: CompensatorCandidate[];
  modelTier?: Tier;
  scopeViolation?: boolean;
  /** When set, tiers that need a capability this path lacks become T4 (D5). */
  captureFidelity?: CaptureFidelity;
}

const READ_HTTP = new Set(["GET", "HEAD", "OPTIONS", "TRACE"]);
const SEARCH_SEGMENTS = new Set(["search", "query", "graphql", "batch", "rpc"]);
const READ_FS = new Set(["open", "stat", "readdir", "readlink"]);
const MESSAGE_TOOLS: Record<string, { tier: Tier; leak: string }> = {
  "chat.postMessage": { tier: "T3", leak: "observers" },
  "slack.chat.postMessage": { tier: "T3", leak: "observers" },
  "discord.createMessage": { tier: "T3", leak: "observers" },
  "telegram.sendMessage": { tier: "T3", leak: "observers" },
};

function asRecord(v: JsonValue | undefined): Record<string, JsonValue> {
  if (v && typeof v === "object" && !Array.isArray(v)) return v as Record<string, JsonValue>;
  return {};
}

function str(v: JsonValue | undefined): string | undefined {
  return typeof v === "string" ? v : undefined;
}

function header(headers: Record<string, string> | undefined, name: string): string | undefined {
  if (!headers) return undefined;
  const want = name.toLowerCase();
  for (const [k, v] of Object.entries(headers)) if (k.toLowerCase() === want) return v;
  return undefined;
}

function lastSegment(url: string): string {
  try {
    const u = new URL(url);
    const parts = u.pathname.split("/").filter(Boolean);
    return (parts[parts.length - 1] ?? "").toLowerCase();
  } catch {
    const parts = url.split("?")[0]!.split("/").filter(Boolean);
    return (parts[parts.length - 1] ?? "").toLowerCase();
  }
}

function hasBody(args: Record<string, JsonValue>): boolean {
  const body = args.body;
  if (body === undefined || body === null) return false;
  if (typeof body === "string") return body.length > 0;
  if (typeof body === "object") return Array.isArray(body) ? body.length > 0 : Object.keys(body).length > 0;
  return true;
}

function bodyId(result: JsonValue | undefined): boolean {
  const r = asRecord(result);
  return typeof r.id === "string" || typeof r.id === "number";
}

function isPrivateHost(host: string, extra: string[] = []): boolean {
  const h = host.toLowerCase().replace(/\.$/, "");
  if (extra.some((x) => x.toLowerCase() === h)) return true;
  if (h === "localhost" || h.endsWith(".localhost")) return true;
  if (h.endsWith(".local") || h.endsWith(".internal")) return true;
  if (h === "127.0.0.1" || h.startsWith("127.")) return true;
  if (h === "::1") return true;
  const m = /^(\d+)\.(\d+)\.(\d+)\.(\d+)$/.exec(h);
  if (m) {
    const a = Number(m[1]);
    const b = Number(m[2]);
    if (a === 10) return true;
    if (a === 172 && b >= 16 && b <= 31) return true;
    if (a === 192 && b === 168) return true;
  }
  if (h.startsWith("fd") || h.startsWith("fe80:")) return true;
  return false;
}

export function classifyLocality(
  target: string | undefined,
  internalHosts: string[] = [],
): { locality: ActionRef["locality"]; tier: Tier; reasons: string[] } {
  if (!target) return { locality: "unknown", tier: "T3", reasons: ["locality_unknown_treated_external"] };
  let host = target;
  try {
    if (target.includes("://")) host = new URL(target).hostname;
  } catch {
    host = target;
  }
  if (isPrivateHost(host, internalHosts)) {
    return { locality: "internal", tier: "T1", reasons: ["locality_internal"] };
  }
  return { locality: "external", tier: "T1", reasons: ["locality_external"] };
}

function httpMethod(action: ActionRef, args: Record<string, JsonValue>): string {
  const m = str(args.method) ?? action.name.split(".")[1] ?? "";
  return m.toUpperCase();
}

function httpUrl(action: ActionRef, args: Record<string, JsonValue>): string {
  return str(args.url) ?? action.target ?? action.name.replace(/^http\.[A-Z]+\./, "https://");
}

export function classifyStructural(
  action: ActionRef,
  args: JsonValue,
  ctx: ClassifyContext = {},
): { tier: Tier; reasons: string[]; in_doubt: boolean; locality: ActionRef["locality"] } {
  const rec = asRecord(args);
  const loc = classifyLocality(action.target ?? str(rec.url) ?? str(rec.path), ctx.internalHosts);
  const reasons: string[] = [];
  let in_doubt = false;
  let tier: Tier = "T4";

  if (action.kind === "shell") {
    reasons.push("shell_unclassifiable");
    return { tier: "T4", reasons, in_doubt: false, locality: loc.locality };
  }

  if (action.kind === "http") {
    const method = httpMethod(action, rec);
    const url = httpUrl(action, rec);
    const override = header(asRecord(rec.headers) as unknown as Record<string, string>, "x-http-method-override")
      ?? header(ctx.resultHeaders, "x-http-method-override");
    const headersObj = asRecord(rec.headers) as unknown as Record<string, string>;
    const ov = header(headersObj, "X-HTTP-Method-Override");
    if (ov || override) {
      tier = "T3";
      reasons.push("method_override");
    } else if (READ_HTTP.has(method) && !hasBody(rec)) {
      tier = "T1";
      reasons.push("http_safe_method");
      if (header(ctx.resultHeaders, "set-cookie")) {
        tier = "T2";
        reasons.push("set_cookie");
      }
    } else if (READ_HTTP.has(method) && hasBody(rec)) {
      tier = "T3";
      reasons.push("get_with_body");
    } else if (method === "POST" && SEARCH_SEGMENTS.has(lastSegment(url))) {
      tier = "T3";
      reasons.push("post_search_like");
    } else if (method === "DELETE") {
      if (ctx.recreateFromPreimage && ctx.preimage && !ctx.preimage.truncated) {
        tier = "T3";
        reasons.push("delete_with_preimage");
      } else {
        tier = "T4";
        reasons.push("delete_without_preimage");
      }
    } else if (method === "PUT" || method === "PATCH") {
      if (ctx.resultStatus === 201 && (header(ctx.resultHeaders, "location") || bodyId(ctx.result))) {
        tier = "T3";
        reasons.push("http_create");
      } else if (ctx.preimage && !ctx.preimage.truncated) {
        tier = loc.locality === "internal" ? "T2" : "T3";
        reasons.push("http_update_with_preimage");
      } else {
        tier = "T4";
        reasons.push("http_update_without_preimage");
      }
    } else if (method === "POST" || method === "PUT") {
      if (ctx.resultStatus !== undefined && ctx.resultStatus >= 400 && !header(ctx.resultHeaders, "location") && !bodyId(ctx.result)) {
        tier = "T1";
        reasons.push("nothing_landed");
      } else if (ctx.resultStatus === 201 || header(ctx.resultHeaders, "location") || bodyId(ctx.result)) {
        tier = "T3";
        reasons.push("http_create");
      } else {
        tier = "T3";
        reasons.push("http_mutation");
      }
    } else {
      tier = "T3";
      reasons.push("http_mutation");
    }

    if ((ctx.transportError || ctx.timeout || (ctx.resultStatus !== undefined && ctx.resultStatus >= 500)) && !READ_HTTP.has(method)) {
      in_doubt = true;
      reasons.push("http_in_doubt");
      if (tier === "T1") {
        tier = method === "DELETE" ? "T4" : "T3";
      }
    }
    if (ctx.resultStatus !== undefined && ctx.resultStatus >= 400 && ctx.resultStatus < 500 && !header(ctx.resultHeaders, "location") && !bodyId(ctx.result)) {
      if (!in_doubt) {
        tier = "T1";
        reasons.push("client_error_nothing_landed");
      }
    }
    if (action.locality === "internal" || loc.locality === "internal") {
      if (tier === "T1" && !(READ_HTTP.has(method) && !hasBody(rec))) {
        tier = "T2";
        reasons.push("internal_write_floor");
      }
    }
    return { tier, reasons, in_doubt, locality: loc.locality };
  }

  if (action.kind === "sql") {
    const sql = str(rec.sql) ?? str(rec.statement) ?? "";
    const parsed = classifySql(sql);
    if (parsed.multiStatement) {
      return { tier: "T4", reasons: ["sql_multi_statement"], in_doubt: false, locality: loc.locality };
    }
    if (parsed.kind === "SELECT" && !parsed.forUpdate && !parsed.modifyingCte) {
      return { tier: "T1", reasons: ["sql_select"], in_doubt: false, locality: loc.locality };
    }
    if (parsed.kind === "SELECT" && parsed.forUpdate) {
      return { tier: "T2", reasons: ["sql_select_for_update"], in_doubt: false, locality: loc.locality };
    }
    const internal = (action.locality ?? loc.locality) === "internal";
    const t23 = (): Tier => (internal ? "T2" : "T3");
    if (parsed.kind === "INSERT") {
      if (ctx.pkResolvable === false) return { tier: "T4", reasons: ["insert_no_pk"], in_doubt: false, locality: loc.locality };
      if (ctx.pkResolvable === true || parsed.returning) return { tier: t23(), reasons: ["insert_pk"], in_doubt: false, locality: loc.locality };
      return { tier: "T4", reasons: ["insert_no_pk"], in_doubt: false, locality: loc.locality };
    }
    if (parsed.kind === "UPDATE") {
      const over = (ctx.affectedRows ?? 0) > (ctx.maxPreimageRows ?? 10_000);
      if (!ctx.preimage || ctx.preimage.truncated || over) {
        return { tier: "T4", reasons: ["update_no_preimage"], in_doubt: false, locality: loc.locality };
      }
      return { tier: t23(), reasons: ["update_preimage"], in_doubt: false, locality: loc.locality };
    }
    if (parsed.kind === "DELETE") {
      if (!ctx.preimage || ctx.preimage.truncated) {
        return { tier: "T4", reasons: ["delete_no_row_images"], in_doubt: false, locality: loc.locality };
      }
      return { tier: t23(), reasons: ["delete_row_images"], in_doubt: false, locality: loc.locality };
    }
    if (["TRUNCATE", "DROP", "ALTER", "CREATE", "GRANT", "REINDEX", "VACUUM"].includes(parsed.kind)) {
      return { tier: "T4", reasons: ["sql_ddl"], in_doubt: false, locality: loc.locality };
    }
    if (parsed.kind === "CALL" || parsed.kind === "EXEC") {
      return { tier: "T4", reasons: ["sql_call"], in_doubt: false, locality: loc.locality };
    }
    return { tier: "T4", reasons: ["sql_unknown"], in_doubt: false, locality: loc.locality };
  }

  if (action.kind === "fs") {
    const op = (str(rec.op) ?? action.name.split(".")[1] ?? "").toLowerCase();
    const path = str(rec.path) ?? str(rec.realpath) ?? action.target ?? "";
    const roots = ctx.writableRoots ?? [];
    if (ctx.realpathEscapes || ctx.networkOrFuseMount) {
      return { tier: "T4", reasons: ["fs_escape_or_remote"], in_doubt: false, locality: "internal" };
    }
    if (roots.length > 0) {
      const real = str(rec.realpath) ?? path;
      const ok = roots.some((r) => real === r || real.startsWith(r.endsWith("/") ? r : r + "/"));
      if (!ok) return { tier: "T4", reasons: ["scope_violation"], in_doubt: false, locality: "internal" };
    }
    if (READ_FS.has(op) || op === "open_rdonly") {
      return { tier: "T1", reasons: ["fs_read"], in_doubt: false, locality: "internal" };
    }
    if (op === "write" || op === "truncate") {
      if (ctx.preimage?.truncated) return { tier: "T4", reasons: ["preimage_truncated"], in_doubt: false, locality: "internal" };
      if (ctx.preimage?.kind === "fs_absent") return { tier: "T2", reasons: ["fs_create"], in_doubt: false, locality: "internal" };
      if (ctx.preimage?.kind === "fs_bytes") return { tier: "T2", reasons: ["fs_overwrite"], in_doubt: false, locality: "internal" };
      if (ctx.preimage === undefined) {
        const fid = ctx.captureFidelity;
        if (fid && fid !== "full") {
          return { tier: "T4", reasons: ["fidelity_downgrade_no_preimage"], in_doubt: false, locality: "internal" };
        }
        return { tier: "T2", reasons: ["fs_write_assume_capture"], in_doubt: false, locality: "internal" };
      }
      return { tier: "T2", reasons: ["fs_write"], in_doubt: false, locality: "internal" };
    }
    if (op === "mkdir" || op === "rename" || op === "chmod" || op === "chown" || op === "utimes") {
      return { tier: "T2", reasons: [`fs_${op}`], in_doubt: false, locality: "internal" };
    }
    if (op === "unlink") {
      if (ctx.preimage && !ctx.preimage.truncated) return { tier: "T2", reasons: ["fs_unlink_preimage"], in_doubt: false, locality: "internal" };
      return { tier: "T4", reasons: ["fs_unlink_no_preimage"], in_doubt: false, locality: "internal" };
    }
    return { tier: "T4", reasons: ["fs_unknown"], in_doubt: false, locality: "internal" };
  }

  if (action.kind === "mcp_tool" || action.kind === "sdk_fn") {
    const tool = str(rec.tool) ?? action.name.split(".").slice(-1)[0] ?? action.name;
    const email = /smtp|resend|ses|postmark|email|sendmail/i.test(action.name + tool);
    const sms = /sms|push_notification|notify\.push/i.test(action.name + tool);
    if (email || sms) return { tier: "T4", reasons: ["message_irreversible"], in_doubt: false, locality: loc.locality };
    for (const [k, v] of Object.entries(MESSAGE_TOOLS)) {
      if (tool === k || action.name.endsWith(k) || action.name.includes(k)) {
        return { tier: v.tier, reasons: ["message_retractable"], in_doubt: false, locality: loc.locality };
      }
    }
    return { tier: "T4", reasons: ["mcp_unknown"], in_doubt: false, locality: loc.locality };
  }

  reasons.push("unknown_action_kind");
  return { tier: "T4", reasons, in_doubt: false, locality: loc.locality };
}

export function joinTier(s: TierEvidence): { tier: Tier; reason: string[] } {
  let tier = maxTier(s.structural, s.locality);
  if (s.manifest && !s.scopeViolation) tier = s.manifest.tier;
  if (s.scopeViolation) tier = "T4";
  if (s.model) tier = maxTier(tier, s.model.tier);
  const reason = [...s.reasons];
  if (!s.compensatorMatched && tier === "T3") {
    return { tier, reason: [...reason, "compensator_unavailable"] };
  }
  return { tier, reason };
}

export function detectScopeViolation(
  action: ActionRef,
  args: JsonValue,
  manifest?: ActionSignature,
  writableRoots?: string[],
): boolean {
  if (action.kind === "fs") {
    const path = str(asRecord(args).path) ?? action.target ?? "";
    if (writableRoots && writableRoots.length > 0) {
      const ok = writableRoots.some((r) => path === r || path.startsWith(r.endsWith("/") ? r : r + "/"));
      if (!ok) return true;
    }
    if (manifest?.match.kind === "fs" && manifest.match.path_glob) {
      const glob = manifest.match.path_glob;
      if (glob.includes("**")) {
        const prefix = glob.split("**")[0] ?? "";
        if (prefix && !path.startsWith(prefix.replace(/\/$/, ""))) return true;
      }
    }
  }
  return false;
}

export function classifyAction(
  action: ActionRef,
  args: JsonValue,
  ctx: ClassifyContext = {},
): ClassificationRecord {
  const structural = classifyStructural(action, args, ctx);
  const loc = classifyLocality(action.target ?? str(asRecord(args).url), ctx.internalHosts);
  const scopeViolation =
    ctx.scopeViolation ?? detectScopeViolation(action, args, ctx.manifest, ctx.writableRoots);
  const locTier: Tier =
    loc.locality === "internal" && structural.tier === "T1" ? "T1" : loc.locality === "internal" ? "T2" : "T1";
  const evidence: TierEvidence = {
    structural: structural.tier,
    locality: locTier,
    manifest: ctx.manifest ? { tier: ctx.manifest.tier } : undefined,
    model: ctx.modelTier ? { tier: ctx.modelTier } : undefined,
    scopeViolation,
    compensatorMatched: ctx.compensatorMatched ?? false,
    reasons: [...structural.reasons, ...loc.reasons],
  };
  const joined = joinTier(evidence);
  const sources: ClassificationSource[] = [
    { source: "structural", tier: structural.tier, reasons: structural.reasons },
    { source: "locality", tier: locTier, reasons: loc.reasons },
  ];
  if (ctx.manifest) sources.push({ source: "manifest", tier: ctx.manifest.tier, reasons: ["manifest"] });
  if (ctx.modelTier) sources.push({ source: "model", tier: ctx.modelTier, reasons: ["model"] });
  let tier = joined.tier;
  const reasons = [...joined.reason];
  if (
    ctx.captureFidelity &&
    ctx.captureFidelity !== "full" &&
    structural.reasons.includes("fidelity_downgrade_no_preimage")
  ) {
    tier = "T4";
    if (!reasons.includes("fidelity_downgrade_no_preimage")) reasons.push("fidelity_downgrade_no_preimage");
  }
  return {
    tier,
    sources,
    reasons,
    candidates: ctx.candidates ?? [],
    scope_violation: scopeViolation,
  };
}

File: sdk-python/src/latticeag_vekrevert/core/taxonomy.py
"""Rule-based hot-path classification. Port of packages/core/src/taxonomy.ts (D8)."""

from __future__ import annotations

from urllib.parse import urlparse

from .sqlkind import classify_sql

TIER_ORDER = {"T1": 1, "T2": 2, "T3": 3, "T4": 4}
READ_HTTP = {"GET", "HEAD", "OPTIONS", "TRACE"}
SEARCH_SEGMENTS = {"search", "query", "graphql", "batch", "rpc"}
READ_FS = {"open", "stat", "readdir", "readlink"}
MESSAGE_TOOLS = {
    "chat.postMessage": "T3",
    "slack.chat.postMessage": "T3",
    "discord.createMessage": "T3",
    "telegram.sendMessage": "T3",
}


def max_tier(a: str, b: str) -> str:
    return a if TIER_ORDER[a] >= TIER_ORDER[b] else b


def join_tier(s: dict) -> dict:
    tier = max_tier(s["structural"], s["locality"])
    if s.get("manifest") and not s.get("scopeViolation"):
        tier = s["manifest"]["tier"]
    if s.get("scopeViolation"):
        tier = "T4"
    if s.get("model"):
        tier = max_tier(tier, s["model"]["tier"])
    reason = list(s.get("reasons") or [])
    if not s.get("compensatorMatched") and tier == "T3":
        return {"tier": tier, "reason": reason + ["compensator_unavailable"]}
    return {"tier": tier, "reason": reason}


def _as_record(v) -> dict:
    return v if isinstance(v, dict) else {}


def _str(v) -> str | None:
    return v if isinstance(v, str) else None


def _header(headers: dict | None, name: str) -> str | None:
    if not headers:
        return None
    want = name.lower()
    for k, v in headers.items():
        if str(k).lower() == want:
            return str(v)
    return None


def _last_segment(url: str) -> str:
    try:
        parts = [p for p in urlparse(url).path.split("/") if p]
        return (parts[-1] if parts else "").lower()
    except Exception:
        parts = [p for p in url.split("?")[0].split("/") if p]
        return (parts[-1] if parts else "").lower()


def _has_body(args: dict) -> bool:
    body = args.get("body")
    if body is None:
        return False
    if isinstance(body, str):
        return len(body) > 0
    if isinstance(body, list):
        return len(body) > 0
    if isinstance(body, dict):
        return len(body) > 0
    return True


def _body_id(result) -> bool:
    r = _as_record(result)
    return isinstance(r.get("id"), (str, int))


def _is_private_host(host: str, extra: list[str] | None = None) -> bool:
    extra = extra or []
    h = host.lower().rstrip(".")
    if any(x.lower() == h for x in extra):
        return True
    if h == "localhost" or h.endswith(".localhost"):
        return True
    if h.endswith(".local") or h.endswith(".internal"):
        return True
    if h == "127.0.0.1" or h.startswith("127."):
        return True
    if h == "::1":
        return True
    m = __import__("re").match(r"^(\d+)\.(\d+)\.(\d+)\.(\d+)$", h)
    if m:
        a, b = int(m.group(1)), int(m.group(2))
        if a == 10:
            return True
        if a == 172 and 16 <= b <= 31:
            return True
        if a == 192 and b == 168:
            return True
    if h.startswith("fd") or h.startswith("fe80:"):
        return True
    return False


def classify_locality(target: str | None, internal_hosts: list[str] | None = None) -> dict:
    internal_hosts = internal_hosts or []
    if not target:
        return {"locality": "unknown", "tier": "T3", "reasons": ["locality_unknown_treated_external"]}
    host = target
    try:
        if "://" in target:
            host = urlparse(target).hostname or target
    except Exception:
        host = target
    if _is_private_host(host, internal_hosts):
        return {"locality": "internal", "tier": "T1", "reasons": ["locality_internal"]}
    return {"locality": "external", "tier": "T1", "reasons": ["locality_external"]}


AGENT_FS_WRITE = {
    "write_file",
    "writefile",
    "str_replace",
    "patch",
    "edit",
    "mcp__filesystem__write_file",
    "mcp__filesystem__edit_file",
}
AGENT_FS_READ = {
    "read_file",
    "readfile",
    "search_files",
    "list_directory",
    "mcp__filesystem__read_text_file",
    "mcp__filesystem__list_directory",
}
AGENT_HTTP_READ = {"web_search", "websearch", "web_extract", "webextract"}


def rewrite_agent_tool(action: dict, args) -> dict | None:
    kind = action.get("kind")
    if kind not in ("mcp_tool", "sdk_fn"):
        return None
    rec = _as_record(args)
    tool = (_str(rec.get("tool")) or (action.get("name") or "").split(".")[-1] or action.get("name") or "").lower()
    path = _str(rec.get("path")) or _str(rec.get("file_path")) or _str(rec.get("target")) or "/eval-sandbox/file"
    if tool in AGENT_FS_WRITE:
        return {
            "action": {"kind": "fs", "name": f"fs.write.{path}", "target": path, "locality": "internal"},
            "args": {**rec, "op": "write", "path": path, "realpath": path},
        }
    if tool in AGENT_FS_READ:
        op = "readdir" if ("search" in tool or "list" in tool) else "stat"
        return {
            "action": {"kind": "fs", "name": f"fs.{op}.{path}", "target": path, "locality": "internal"},
            "args": {**rec, "op": op, "path": path, "realpath": path},
        }
    if tool in AGENT_HTTP_READ:
        return {
            "action": {
                "kind": "http",
                "name": "http.GET.search.invalid/search",
                "target": "search.invalid",
                "locality": "external",
            },
            "args": {"method": "GET", "url": _str(rec.get("url")) or "https://search.invalid/search"},
        }
    return None


def classify_structural(action: dict, args, ctx: dict | None = None) -> dict:
    ctx = ctx or {}
    aliased = rewrite_agent_tool(action, args)
    if aliased:
        return classify_structural(aliased["action"], aliased["args"], ctx)
    rec = _as_record(args)
    loc = classify_locality(action.get("target") or _str(rec.get("url")) or _str(rec.get("path")), ctx.get("internalHosts"))
    kind = action.get("kind")
    if kind == "shell":
        return {"tier": "T4", "reasons": ["shell_unclassifiable"], "in_doubt": False, "locality": loc["locality"]}
    if kind == "http":
        method = (_str(rec.get("method")) or "")
        if not method:
            parts = (action.get("name") or "").split(".")
            method = parts[1] if len(parts) > 1 else ""
        method = method.upper()
        url = _str(rec.get("url")) or action.get("target") or ""
        headers_obj = _as_record(rec.get("headers"))
        ov = _header(headers_obj, "X-HTTP-Method-Override") or _header(ctx.get("resultHeaders"), "x-http-method-override")
        reasons: list[str] = []
        in_doubt = False
        tier = "T4"
        if ov:
            tier, reasons = "T3", ["method_override"]
        elif method in READ_HTTP and not _has_body(rec):
            tier, reasons = "T1", ["http_safe_method"]
            if _header(ctx.get("resultHeaders"), "set-cookie"):
                tier, reasons = "T2", reasons + ["set_cookie"]
        elif method in READ_HTTP and _has_body(rec):
            tier, reasons = "T3", ["get_with_body"]
        elif method == "POST" and _last_segment(url) in SEARCH_SEGMENTS:
            tier, reasons = "T3", ["post_search_like"]
        elif method == "DELETE":
            if ctx.get("recreateFromPreimage") and ctx.get("preimage") and not ctx["preimage"].get("truncated"):
                tier, reasons = "T3", ["delete_with_preimage"]
            else:
                tier, reasons = "T4", ["delete_without_preimage"]
        elif method in ("PUT", "PATCH"):
            if ctx.get("resultStatus") == 201 and (_header(ctx.get("resultHeaders"), "location") or _body_id(ctx.get("result"))):
                tier, reasons = "T3", ["http_create"]
            elif ctx.get("preimage") and not ctx["preimage"].get("truncated"):
                tier = "T2" if loc["locality"] == "internal" else "T3"
                reasons = ["http_update_with_preimage"]
            else:
                tier, reasons = "T4", ["http_update_without_preimage"]
        elif method in ("POST", "PUT"):
            st = ctx.get("resultStatus")
            if st is not None and st >= 400 and not _header(ctx.get("resultHeaders"), "location") and not _body_id(ctx.get("result")):
                tier, reasons = "T1", ["nothing_landed"]
            elif st == 201 or _header(ctx.get("resultHeaders"), "location") or _body_id(ctx.get("result")):
                tier, reasons = "T3", ["http_create"]
            else:
                tier, reasons = "T3", ["http_mutation"]
        else:
            tier, reasons = "T3", ["http_mutation"]
        if (ctx.get("transportError") or ctx.get("timeout") or (ctx.get("resultStatus") is not None and ctx["resultStatus"] >= 500)) and method not in READ_HTTP:
            in_doubt = True
            reasons.append("http_in_doubt")
            if tier == "T1":
                tier = "T4" if method == "DELETE" else "T3"
        return {"tier": tier, "reasons": reasons, "in_doubt": in_doubt, "locality": loc["locality"]}
    if kind == "sql":
        sql = _str(rec.get("sql")) or _str(rec.get("statement")) or ""
        parsed = classify_sql(sql)
        if parsed["multiStatement"]:
            return {"tier": "T4", "reasons": ["sql_multi_statement"], "in_doubt": False, "locality": loc["locality"]}
        if parsed["kind"] == "SELECT" and not parsed["forUpdate"] and not parsed["modifyingCte"]:
            return {"tier": "T1", "reasons": ["sql_select"], "in_doubt": False, "locality": loc["locality"]}
        internal = (action.get("locality") or loc["locality"]) == "internal"
        t23 = "T2" if internal else "T3"
        if parsed["kind"] == "INSERT":
            if ctx.get("pkResolvable") is False:
                return {"tier": "T4", "reasons": ["insert_no_pk"], "in_doubt": False, "locality": loc["locality"]}
            if ctx.get("pkResolvable") is True or parsed["returning"]:
                return {"tier": t23, "reasons": ["insert_pk"], "in_doubt": False, "locality": loc["locality"]}
            return {"tier": "T4", "reasons": ["insert_no_pk"], "in_doubt": False, "locality": loc["locality"]}
        if parsed["kind"] == "UPDATE":
            if not ctx.get("preimage") or ctx["preimage"].get("truncated"):
                return {"tier": "T4", "reasons": ["update_no_preimage"], "in_doubt": False, "locality": loc["locality"]}
            return {"tier": t23, "reasons": ["update_preimage"], "in_doubt": False, "locality": loc["locality"]}
        if parsed["kind"] == "DELETE":
            if not ctx.get("preimage") or ctx["preimage"].get("truncated"):
                return {"tier": "T4", "reasons": ["delete_no_row_images"], "in_doubt": False, "locality": loc["locality"]}
            return {"tier": t23, "reasons": ["delete_row_images"], "in_doubt": False, "locality": loc["locality"]}
        return {"tier": "T4", "reasons": ["sql_ddl"], "in_doubt": False, "locality": loc["locality"]}
    if kind == "fs":
        op = (_str(rec.get("op")) or (action.get("name") or "fs.write").split(".")[1]).lower()
        path = _str(rec.get("path")) or _str(rec.get("realpath")) or action.get("target") or ""
        roots = ctx.get("writableRoots") or []
        if ctx.get("realpathEscapes") or ctx.get("networkOrFuseMount"):
            return {"tier": "T4", "reasons": ["fs_escape_or_remote"], "in_doubt": False, "locality": "internal"}
        if roots:
            real = _str(rec.get("realpath")) or path
            ok = any(real == r or real.startswith(r if r.endswith("/") else r + "/") for r in roots)
            if not ok:
                return {"tier": "T4", "reasons": ["scope_violation"], "in_doubt": False, "locality": "internal"}
        if op in READ_FS or op == "open_rdonly":
            return {"tier": "T1", "reasons": ["fs_read"], "in_doubt": False, "locality": "internal"}
        if op in ("write", "truncate"):
            pre = ctx.get("preimage")
            fid = ctx.get("captureFidelity")
            if pre and pre.get("truncated"):
                return {"tier": "T4", "reasons": ["preimage_truncated"], "in_doubt": False, "locality": "internal"}
            if pre and pre.get("kind") == "fs_absent":
                return {"tier": "T2", "reasons": ["fs_create"], "in_doubt": False, "locality": "internal"}
            if pre and pre.get("kind") == "fs_bytes":
                return {"tier": "T2", "reasons": ["fs_overwrite"], "in_doubt": False, "locality": "internal"}
            if pre is None and fid and fid != "full":
                return {"tier": "T4", "reasons": ["fidelity_downgrade_no_preimage"], "in_doubt": False, "locality": "internal"}
            return {"tier": "T2", "reasons": ["fs_write_assume_capture" if pre is None else "fs_write"], "in_doubt": False, "locality": "internal"}
        if op in ("mkdir", "rename", "chmod", "chown", "utimes"):
            return {"tier": f"T2", "reasons": [f"fs_{op}"], "in_doubt": False, "locality": "internal"}
        if op == "unlink":
            if ctx.get("preimage") and not ctx["preimage"].get("truncated"):
                return {"tier": "T2", "reasons": ["fs_unlink_preimage"], "in_doubt": False, "locality": "internal"}
            return {"tier": "T4", "reasons": ["fs_unlink_no_preimage"], "in_doubt": False, "locality": "internal"}
        return {"tier": "T4", "reasons": ["fs_unknown"], "in_doubt": False, "locality": "internal"}
    if kind in ("mcp_tool", "sdk_fn"):
        tool = _str(rec.get("tool")) or (action.get("name") or "").split(".")[-1]
        blob = (action.get("name") or "") + (tool or "")
        if __import__("re").search(r"smtp|resend|ses|postmark|email|sendmail", blob, __import__("re").I):
            return {"tier": "T4", "reasons": ["message_irreversible"], "in_doubt": False, "locality": loc["locality"]}
        if __import__("re").search(r"sms|push_notification|notify\.push", blob, __import__("re").I):
            return {"tier": "T4", "reasons": ["message_irreversible"], "in_doubt": False, "locality": loc["locality"]}
        for k, t in MESSAGE_TOOLS.items():
            if tool == k or (action.get("name") or "").endswith(k) or k in (action.get("name") or ""):
                return {"tier": t, "reasons": ["message_retractable"], "in_doubt": False, "locality": loc["locality"]}
        return {"tier": "T4", "reasons": ["mcp_unknown"], "in_doubt": False, "locality": loc["locality"]}
    return {"tier": "T4", "reasons": ["unknown_action_kind"], "in_doubt": False, "locality": loc["locality"]}


def detect_scope_violation(action: dict, args, manifest=None, writable_roots=None) -> bool:
    if action.get("kind") != "fs":
        return False
    path = _str(_as_record(args).get("path")) or action.get("target") or ""
    if writable_roots:
        ok = any(path == r or path.startswith(r if r.endswith("/") else r + "/") for r in writable_roots)
        if not ok:
            return True
    return False


def classify_action(action: dict, args, ctx: dict | None = None) -> dict:
    ctx = ctx or {}
    structural = classify_structural(action, args, ctx)
    loc = classify_locality(action.get("target") or _str(_as_record(args).get("url")), ctx.get("internalHosts"))
    scope = ctx.get("scopeViolation")
    if scope is None:
        scope = detect_scope_violation(action, args, ctx.get("manifest"), ctx.get("writableRoots"))
    loc_tier = "T1" if loc["locality"] != "internal" or structural["tier"] == "T1" else "T2"
    evidence = {
        "structural": structural["tier"],
        "locality": loc_tier,
        "manifest": {"tier": ctx["manifest"]["tier"]} if ctx.get("manifest") else None,
        "model": {"tier": ctx["modelTier"]} if ctx.get("modelTier") else None,
        "scopeViolation": scope,
        "compensatorMatched": ctx.get("compensatorMatched") or False,
        "reasons": structural["reasons"] + loc["reasons"],
    }
    joined = join_tier(evidence)
    sources = [
        {"source": "structural", "tier": structural["tier"], "reasons": structural["reasons"]},
        {"source": "locality", "tier": loc_tier, "reasons": loc["reasons"]},
    ]
    return {
        "tier": joined["tier"],
        "sources": sources,
        "reasons": joined["reason"],
        "candidates": ctx.get("candidates") or [],
        "scope_violation": scope,
    }
